lib/linter.js

```diff
@@ -898,12 +905,13 @@ class Linter {
         const ecmaVersion = this.currentConfig.parserOptions.ecmaVersion || 5;
 
         // gather scope data that may be needed by the rules
-        this.scopeManager = eslintScope.analyze(sourceCode.ast, {
+        this.scopeManager = this.sourceCode.scopeManager || eslintScope.analyze(sourceCode.ast, {
             ignoreEval: true,
             nodejsScope: ecmaFeatures.globalReturn,
             impliedStrict: ecmaFeatures.impliedStrict,
             ecmaVersion,
             sourceType: this.currentConfig.parserOptions.sourceType || "script",
+            childVisitorKeys: this.sourceCode.visitorKeys || Traverser.DEFAULT_VISITOR_KEYS,
```

Comment:
Would it be better for `sourceCode.visitorKeys` to always default to `Traverser.DEFAULT_VISITOR_KEYS` in the `SourceCode` constructor? That way, `sourceCode.visitorKeys` would always be defined.